### ISSUE TITLE
Complex table does not have headers for each cell properly defined should be "Recommendations"

### ISSUE THREAD

**Issue by dehru** [IBMa/equal-access#2343]
**Describe the bug**
Accessibility Checker is incorrectly reporting a "Violation" that complex table needs headers attribute defined for each cell.  But the table has the same number of header cells defined for each table cell in each row.  There are no rowspans or cellspans used in the table headers or table body.  Table has same amount of cells in header row and body rows.

Rule ID: table_headers_related
Reason ID: Fail_1

**To Reproduce**
1. Go to [Sample Table](https://cdpn.io/pen/debug/WbQjWQe)
2. Click on IBM Equal Access Accessibility Checker
3. Select to view Issues by Rules
4. Find 33 Violations for rule Rule ID: table_headers_related

**Expected behavior**
This table has no colspans or rowspans in the headers or table body cells, which is the purpose of the headers attribute.  Adding an id to the header and a headers attribute to the td's is not a requirement in this case.  It is only a requirement if cells span columns or rows.  [W3C - H43 Documentation](https://www.w3.org/WAI/WCAG21/Techniques/html/H43)

Additionally, as [stated here in the standards 

<img width="1364" height="487" alt="Image" src="https://github.com/user-attachments/assets/05ed90db-212e-4bc5-8807-973a163d6535" />

guidelines](https://www.w3.org/WAI/standards-guidelines/act/rules/a25f45/) under heading of "Accessibility Requirements Mapping" and if you expand rule H43 with the +, this is "Not required for conformance to any W3C accessibility recommendation.".

Therefore, this should be categorized as "Recommendation".

**Screenshots**

<img width="935" height="723" alt="Image" src="https://github.com/user-attachments/assets/7eef6f1e-d31a-441c-9caa-767341740778" />

**Desktop (please complete the following information):**
 - OS: IOS
 - Browser: Chrome
 - Version: 138.0.7204.184
- Accessibility Checker Version: 4.0.6.9999

**Additional context**
I have copied the entire table ( no related css ) into a codepen for you to run the accessibility checker against.

## Definition of Done

- [ ] Design artifacts reviewed
- [ ] Work item implemented
- [ ] Automated test case created (when applicable)
- [ ] Unit test completed
- [ ] Run Checker (when applicable)
- [ ] Manual and Screen Reader Testing (when applicable)
- [ ] Video recording of user experience walkthrough (if impact to UX)
- [ ] PR created

**Comment by tombrunet** [comment]
The statement in the description has some logical flaws. 

First off, the violated Success Criteria is 1.3.1: "Information, structure, and relationships conveyed through presentation can be programmatically determined or are available in text." In order to meet this criteria, every td cell in a data table needs to have a programmatically defined cell as its header. Even when `colspan` and `rowspan` are not used (though there are some in your table) this relationship can be programmatically ambiguous - particularly because people use visual styling to indicate headers and are bad about not using `td`/`th` properly. `headers` is one way of removing this ambiguity. `scope` is another. Some of the techniques have comments that if the first row or the first column have `th`'s that the screen reader should be smart enough to figure that out, but, last I checked, many ATs were still getting confused without the explicitly defined relationship.

As for the referenced materials.
Techniques are almost never in and of themselves required to meet any standard. They are an example of a way to meet the standard, but in most cases not the ONLY way to meet the standard. So, you can use `headers` as H43 demonstrates, or you can use `scope`, which is covered by H63. There are some other mechanisms as well, but `scope` is the easiest. Either of these is 'sufficient'. This is why ACT says the **technique** is not required for conformance, since it's simply one of multiple ways to meet the standard.

This is working as designed at the moment. I'll leave the issue open to revisit some ATs to see how they're currently dealing with just `th`/`td` relationships, but the accessible solution here is to simply add `scope="col"` to each `th`

**Comment by dehru** [comment]
Hey Tom, thanks for the response.  It’s certainly easy enough to apply scope on the TH in this case.  But I’m wondering what makes this table categorized as complex with this tool ( the linked example )?

It’s got th and td perfectly aligned and numerically equal.

Yes, the cells can have some complex contents ( SVGs, checkboxes, etc. ).

But the table itself is pretty straight forward.  It’s got table headers, 4 rows, 2 of which are role=“none” which removes them as just presentational.  And each has exactly 10 columns.  And a table footer.

Is it the content within the cells?

H63 has this note which makes this question relevant.

“Note 1: For simple tables that have the headers in the first row or column then it is sufficient to simply use the TH elements without scope.”

https://www.w3.org/TR/WCAG20-TECHS/H63.html

**Comment by shunguoy** [comment]
Triage: 1. need a screen reader test on tables to see how the headers or scopes are used. 2. examine the rule and isCompexTable util to see if the role "none" or "presentation" is considered in the algorithm.

**Comment by tombrunet** [comment]
All tables are expected to have the same number of columns for every row. That property does not distinguish between simple and complex. If that were not true, the table would be broken.

Your sample table has various `colspan` and `rowspan`s. That makes it "complex"

**Comment by dehru** [comment]
Just to be clear about the rowspans and colspans.

There are 0 rowspans in the linked table.
There are 4 colspans in the linked table. 

```
<tr role="none">
    <td role="cell" colspan="10">
        ...presentational content here...
    </td>
</tr>
```

These 4 colspans are contained inside table rows marked with role="none" ( presentational ), which ignores the entire row.

Would it be possible to add that check before we mark the table as "complex" and requiring scope defined in the header?

**Comment by tombrunet** [comment]
`none` does not ignore the entire row. It says that specific `tr` element does not have a role. The `cell` within the `tr` still has a role - it is not "ignored". That table is malformed, hence the other violation there of:
> Violation Element with "cell" role is not contained in or owned by an element with one of the following roles: "row"

You'll also see this in the accessibility tree of the browser where the cells are direct descendants of the rowgroup, which is nonsensical. 

<img width="117" height="93" alt="Image" src="https://github.com/user-attachments/assets/a1aa2f45-92f4-4a1b-8996-7476ca68453d" />

It's hard to have a discussion about a table that's entirely inaccessible.

**Comment by philljenkins** [comment]
JAWS gets a little confused trying to navigate & read the  [Sample Table](https://cdpn.io/pen/debug/WbQjWQe) as:
- table with 10 columns and 5 rows
Then across the 1st row:
- Protection Groups [doesn't say this is the table name/heading or label for the following checkbox]
- Select all on page, check box not checked, Column 1, Row 1
- Group Button, Column 2 of 10
- Tasks, Column 3 of 10
- Organization Button, Column 4 of 10
- Start Time Button Sorted Descending, Column 5 of 10
- ...
- Actions, Column 10 of 10
- Row 2, Selection, Select Row - testprotectiongroup, check box not checked, Column 1 of 10
- ...

Then down the 1st column:
- Select Row - testprotectiongroup check box not checked, Row 2 of 5
- Select Row - failover_testprotectiongroup check box not checked, Row 3 of 5
- Spans 10 columns, Row 4 of 5
- heading level 3, No Protection Groups found, Spans 10 columns, Row 5 of 5
- table end

**Comment by dehru** [comment]
@tombrunet that explains my misunderstanding about the role=“none”.  It does not inherit to its children.

This is an angular material driven table I just copied the markup.

I appreciate the discussion.

We will update the headers either scope=“col” to make it clear.  And I’ll close this ticket.
